packages/simcore-sdk/src/simcore_sdk/node_data/data_manager.py

```diff
@@ -1,222 +1,260 @@
+    node_uuid: NodeID,
+    path: Path,
+    *,
+    is_archive: bool,
 ) -> bool:
     """
     :returns True if an entry is present inside the files_metadata else False
     """
-    s3_object = _create_s3_object(project_id, node_uuid, _get_archive_name(file_path))
-    log.debug("Checking if s3_object='%s' is present", s3_object)
+    s3_object = __create_s3_object_key(
+        project_id, node_uuid, __get_s3_name(path, is_archive=is_archive)
+    )
+    _logger.debug("Checking if s3_object='%s' is present", s3_object)
     return await filemanager.entry_exists(
         user_id=user_id,
         store_id=SIMCORE_LOCATION,
         s3_object=s3_object,
+        is_directory=not is_archive,
+    )
+
+
+async def _delete_legacy_archive(
+    user_id: UserID, project_id: ProjectID, node_uuid: NodeID, path: Path
+) -> None:
+    """removes the .zip state archive from storage"""
+    s3_object = __create_s3_object_key(
+        project_id, node_uuid, __get_s3_name(path, is_archive=True)
+    )
+    _logger.debug("Deleting s3_object='%s' is archive", s3_object)
+    await filemanager.delete_file(
+        user_id=user_id, store_id=SIMCORE_LOCATION, s3_object=s3_object
+    )
+
+
+async def push(
+    user_id: UserID,
+    project_id: ProjectID,
+    node_uuid: NodeID,
+    source_path: Path,
+    *,
+    io_log_redirect_cb: LogRedirectCB,
+    r_clone_settings: RCloneSettings,
+    exclude_patterns: set[str] | None = None,
+    progress_bar: ProgressBarData,
+) -> None:
+    """pushes and removes the legacy archive if present"""
+
+    await _push_directory(
+        user_id=user_id,
+        project_id=project_id,
+        node_uuid=node_uuid,
+        source_path=source_path,
+        r_clone_settings=r_clone_settings,
+        exclude_patterns=exclude_patterns,
+        io_log_redirect_cb=io_log_redirect_cb,
+        progress_bar=progress_bar,
+    )
+    archive_exists = await _state_metadata_entry_exists(
+        user_id=user_id,
+        project_id=project_id,
+        node_uuid=node_uuid,
+        path=source_path,
+        is_archive=True,
     )
+
+    if not archive_exists:
+        return
+
+    with log_context(_logger, logging.INFO, "removing legacy data archive"):
+        await _delete_legacy_archive(
+            user_id=user_id,
+            project_id=project_id,
+            node_uuid=node_uuid,
+            path=source_path,
+        )
+
+
+async def pull(
+    user_id: UserID,
+    project_id: ProjectID,
+    node_uuid: NodeID,
+    destination_path: Path,
+    *,
+    io_log_redirect_cb: LogRedirectCB,
+    r_clone_settings: RCloneSettings,
+    progress_bar: ProgressBarData,
+) -> None:
+    """restores the state folder"""
+
+    state_archive_exists = await _state_metadata_entry_exists(
+        user_id=user_id,
+        project_id=project_id,
+        node_uuid=node_uuid,
+        path=destination_path,
+        is_archive=True,
+    )
+    state_directory_exists = await _state_metadata_entry_exists(

```

Comment:
why do you do this call if the first one is already True? These are wasting I/O calls.

I would suggest to do first the legacy call, if yes download it.
later do the modern call.

Or better would be:
1. check for folder
2. if yes, sync with new mode
3. if no, check for archive
4. pull the archive

Reply:
refactored it